Cleanup containers

JustinDrake opened the pull request:
1) Use custom types in container definitions (fix #1153)
2) Renamings

* `deposit_index` => `eth1_deposit_index` (more precise and future proof; consistent with `eth1_data` and `eth1_data_votes` in "Eth1" grouping)
* <s>`previous_epoch_attestations` => `previous_attestations` (consistent with `previous_crosslinks`, `previous_justified_epoch` and `previous_justified_root`; shorter)</s>
* `validator_registry` => `validators` (cleaner; more semantically correct—the registry is now *both* `validators` and `balances`; shorter)
* Remove most `latest_` prefixes (cleaner; shorter; the prefix does not add much other than clutter; we make an exception for `latest_block_header`)
3) Add description to not-immediately-obvious container fields
4) Semantically group `BeaconState` fields, following a logical progression (see item 14 in #1054)

```python
    # Versioning
    genesis_time: uint64
    slot: Slot
    fork: Fork
    # History
    latest_block_header: BeaconBlockHeader
    block_roots: Vector[Hash, SLOTS_PER_HISTORICAL_ROOT]
    state_roots: Vector[Hash, SLOTS_PER_HISTORICAL_ROOT]
    historical_roots: List[Hash]
    # Eth1
    eth1_data: Eth1Data
    eth1_data_votes: List[Eth1Data]
    eth1_deposit_index: uint64
    # Registry
    validators: List[Validator]
    balances: List[Gwei]
    # Shuffling
    start_shard: Shard
    randao_mixes: Vector[Hash, RANDAO_MIXES_LENGTH]
    active_index_roots: Vector[Hash, ACTIVE_INDEX_ROOTS_LENGTH]
    # Slashings
    slashed_balances: Vector[Gwei, SLASHED_EXIT_LENGTH]
    # Attestations
    previous_epoch_attestations: List[PendingAttestation]
    current_epoch_attestations: List[PendingAttestation]
    # Crosslinks
    previous_crosslinks: Vector[Crosslink, SHARD_COUNT]
    current_crosslinks: Vector[Crosslink, SHARD_COUNT]
    # Justification
    previous_justified_epoch: Epoch
    previous_justified_root: Hash
    current_justified_epoch: Epoch
    current_justified_root: Hash
    justification_bitfield: uint64
    # Finality
    finalized_epoch: Epoch
    finalized_root: Hash
```

djrtwo commented:
> validator_registry => validators

We debated this long ago and decided that `validator_registry` was more explicit -- it is the entire registry of validators rather than just a list of validators

djrtwo commented:
> previous_epoch_attestations => previous_attestations (consistent with previous_crosslinks, previous_justified_epoch and previous_justified_root; shorter)

There is a difference between `previous_epoch_attestations` and `previous_crosslinks`/`previous_justified_epoch`. `previous_epoch_attestations` are the attestations _for_ the previous epoch. Whereas `previous_crosslinks` were the `crosslinks` _during_ the previous epoch. Same with `previous_justified_epoch` -- this was the latest `justified_epoch` _during_ the previous epoch.

JustinDrake commented:
> We debated this long ago and decided that `validator_registry` was more explicit

I was in favour of `validator_registry` in that debate. Since then the situation has changed with the registry split being into two parts:

```
    # Registry
    validators: List[Validator]
    balances: List[Gwei]
```

djrtwo commented on specs/core/0_beacon-chain.md:
```diff
@@ -528,55 +488,45 @@ class BeaconBlock(Container):
+    slot: Slot
+    fork: Fork
+    # History
+    parent_block_header: BeaconBlockHeader
```
Not sure about the semantics here.
The `latest_block_header` seen _should_ be the `parent` of the next block, but it is not a parent when recorded and not during any skipped slots when it is recorded into `block_roots`.

JustinDrake replied:
Agreed that naming is awkward here. Brainstorming:

* `block_header` (matches the type nicely; vagueness hides difficult-to-explain subtleties)
* `partial_block_header` (a bit more accurate)
* `latest_block_header` (make an exception here for "latest" prefix)
* `latest_partial_block_header` (a bit of mouthful)

djrtwo replied:
`latest_` seems reasonable and purposefully descriptive here (as opposed to it just being thrown on all the history arrays).

djrtwo replied:
bump @JustinDrake 

Just handled merge conflict. let's resolve this and get merged.

JustinDrake commented on specs/core/0_fork-choice.md:
```diff
@@ -68,8 +68,8 @@ def get_ancestor(store: Store, block: BeaconBlock, slot: Slot) -> BeaconBlock:
 
-* Let `get_latest_attestation(store: Store, index: ValidatorIndex) -> Attestation` be the attestation with the highest slot number in `store` from the validator with the given `index`. If several such attestations exist, use the one the validator `v` observed first.
-* Let `get_latest_attestation_target(store: Store, index: ValidatorIndex) -> BeaconBlock` be the target block in the attestation `get_latest_attestation(store, index)`.
+* Let `get_attestation(store: Store, index: ValidatorIndex) -> Attestation` be the attestation with the highest slot number in `store` from the validator with the given `index`. If several such attestations exist, use the one the validator `v` observed first.
```
Looks like my find-and-replace was a bit too aggressive

JustinDrake commented:
@protolambda, @hwwhww Looks like there is a race condition in the linter CI. I didn't change phase 1 yet the CI is complaining about it.

@djrtwo I think I've addressed all issues, and I've added a bunch of descriptions to field containers. Feel free to edit/add/remove those as you feel appropriate :)

protolambda commented:
@JustinDrake I'll take a closer look. Best guess would be that you changed something in phase 0, and it wound up as part of phase 1.

protolambda commented:
@JustinDrake fixed linting in 75b4692 . Not a race condition, just one missing space. You can run `make -B pyspec` and `make lint` to check your changes locally. :+1:
